### PR TITLE
[DEV-3970] Automating deployment

### PR DESCRIPTION
**Pull request by namack75**

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            steps {
+                sh 'make all'
+            }
+        }
+        stage('Upload - Development') {
+            when {
+                environment name: 'RCH_ENVIRONMENT', value: 'development'
+            }
+            steps {
+                withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'jenkins-aws-development', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
```
Comment (namack75) [inline Jenkinsfile]: This allows us to NOT have to map a volume to .aws and root users.

I have added the creds to Jenkins.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+RUN apt-get update \
+    && apt-get install -y \
+        python3-pip \
+    && pip3 install --upgrade pip \
+    && apt-get clean
+
+RUN pip3 --no-cache-dir install --upgrade awscli boto boto3
+
+ENV HOME .
```
Comment (namack75) [inline Dockerfile]: Binds the Home dir to current work dir, by pass need to pass in Root User.

File: Jenkinsfile (hunk 1 above)
Comment (cjsedwards) [reply Jenkinsfile]: Is there any way to restrict access to this in Jenkins? i.e. we don't want to have just anyone log into jenkins and see these creds?

File: Dockerfile (hunk 2 above)
Comment (namack75) [reply Dockerfile]: Moved to Jenkinsfile

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,66 @@
+pipeline {
+    agent {
+        dockerfile {
+        }
+    }
+    environment {
+        HOME = '.'
```
Comment (namack75) [inline Jenkinsfile]: Binds the Home dir to current work dir, by pass need to pass in Root User.

File: package-lock.json [hunk 4]
```diff
@@ -1,7513 +0,0 @@
-{
```
Comment (landonboyd) [inline package-lock.json]: Are you sure we want to delete this?  I thought it was meant to be under version control.

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                    sh 'aws s3 mv ./dist/reach.min.js s3://assets.rch.red/sdk/checkout-web/reach.min.js'
+                }
+            }
+        }
+        stage('Upload - Sandbox') {
+            when {
+                environment name: 'RCH_ENVIRONMENT', value: 'sandbox'
+            }
+            steps {
+                withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'jenkins-aws-sandbox', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
```
Comment (cjsedwards) [inline Jenkinsfile]: This feels like a lot of duplication.  Is there not a way to set the credentials that we're passing in at the jenkins project level?

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    sh 'aws s3 mv ./dist/reach.min.js s3://assets.rch.how/sdk/checkout-web/reach.min.js'
```
Comment (cjsedwards) [inline Jenkinsfile]: Could we also make the base URL a jenkins job parameter? or make the base_url keyd off of RCH_ENVIRONMENT?

File: package-lock.json (hunk 4 above)
Comment (landonboyd) [reply package-lock.json]: This would presumably allow you to have repeatable builds for any point in time.

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                }
+            }
+        }
+        stage('Upload - Production') {
+            when {
+                environment name: 'RCH_ENVIRONMENT', value: 'production'
+            }
+            steps {
+                withCredentials([[$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: 'jenkins-aws-production', secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
+                    sh 'aws s3 mv ./dist/reach.min.js s3://assets.rch.io/sdk/checkout-web/reach.min.js'
```
Comment (landonboyd) [inline Jenkinsfile]: As discussed, the s3 bucket should be parameterized since this is a public repo.

File: Jenkinsfile (hunk 1 above)
Comment (namack75) [reply Jenkinsfile]: believe the creds are encrypted once entered into the credentials. so no chance of viewing.

File: Jenkinsfile (hunk 5 above)
Comment (namack75) [reply Jenkinsfile]: maybe.

Comment (namack75) [reply Jenkinsfile]: Done

File: Jenkinsfile (hunk 6 above)
Comment (namack75) [reply Jenkinsfile]: Ive changed it to an environment parameter now.

File: Jenkinsfile (hunk 7 above)
Comment (namack75) [reply Jenkinsfile]: done

File: package-lock.json (hunk 4 above)
Comment (namack75) [reply package-lock.json]: I noticed that it was rebuilt on first npm, and broom does a clean install each time.

Comment (namack75) [reply package-lock.json]: re-added from dev branch

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        HOME = '.'
+    }
+    stages {
+        stage('Setup') {
+            steps {
+                sh 'npm --version'
+                sh 'pwd'
+                sh 'env'
+            }
+        }
```
[the remark this answers is not in the document]
Comment (namack75) [reply Jenkinsfile]: The jenkins job will fail, and show red.

**Pull request merged**